### ISSUE TITLE
[Proposal] Smart Enums In C# Like Java

### ISSUE THREAD

**Issue by ProgrammingFire** [dotnet/csharplang#5936]
### Discussed in https://github.com/dotnet/csharplang/discussions/5935

<div type='discussions-op-text'>

<sup>Originally posted by **ProgrammingFire** March 21, 2022</sup>
This proposal is about `enums` which is a pretty useful type in object-oriented programming but the thing that is missing from enum in C# is a feature from java that enums can be like classes a bit like they can have special properties, and have constructors and more...

An example for it will be a simple enum `Subscription`

In Java:
```java
public enum Subscription {
    FREE("Free Tier"),
    PRIME("Premium Tier"),
    VIP("VIP Tier");

    private final String name;

    public Subscription(String name) {
        this.name = name;
    }

    public String getName() {
        return this.name;
    }
}
```

In C#:
```csharp
public enum Subscription {
    Free,
    Prime,
    Vip;
}
```

In C# I am not able to add constructors, properties, and methods. although there is a library to do this [`SmartEnum`](https://github.com/Ardalis/SmartEnum) but wouldn't be nice if we have this feature built-in to C# where we could do something like this

```csharp
public enum Subscription {
    Free("Free Tier"),
    Prime("Premium Tier"),
    Vip("VIP Tier");

    private const string Name;

    public Subscription(string bame) {
        Name = name;
    }

    public string GetName()
    {
        return Name
    }
}
```</div>

**Comment by brianebeling** [comment]
While I think that works and doesn't look half bad, what is the reason you cannot use a combination of a repository and an enum or any arbitrary value as key, that returns some record.
Why the record? Well, it's immutable and has value equality while being a reference type.

That could look like: `var subscription = repository.Get(Subscriptions.Free)` and then I can call all the method you specified on my subscription record.

**Comment by HaloFour** [comment]
Oracles holds a patent on the implementation of enums in Java.  IMO C# would do better to adopt Swift's version of enums instead which can also serve as discriminated unions.

**Comment by bernd5** [comment]
Why not using a [record](https://dotnetfiddle.net/lohY20)?

E.g.:
```csharp
using System;

var free = Subscription.FREE;
Console.WriteLine(free == Subscription.FREE);

public record struct Subscription {
    public static readonly Subscription FREE = new("Free Tier", 0);
    public static readonly Subscription PRIME = new("Premium Tier", 25);
    public static readonly Subscription VIP = new("VIP Tier", 50);
    
    public String Name { get; }
    public int Discount { get; }
    
    public Subscription(String name, int discount) {
        Name = name;
        Discount = discount;
    }
}
```

**Comment by ProgrammingFire** [comment]
> Why not using a [record](https://dotnetfiddle.net/lohY20)?
> 
> E.g.:
> 
> ```cs
> using System;
> 
> var free = Subscription.FREE;
> Console.WriteLine(free == Subscription.FREE);
> 
> public record struct Subscription {
>     public static readonly Subscription FREE = new("Free Tier", 0);
>     public static readonly Subscription PRIME = new("Premium Tier", 25);
>     public static readonly Subscription VIP = new("VIP Tier", 50);
>     
>     public String Name { get; }
>     public int Discount { get; }
>     
>     public Subscription(String name, int discount) {
>         Name = name;
>         Discount = discount;
>     }
> }
> ```

`Enums` Code Look More Clear And Readable In Comparison With `Record` Or `Classes`

**Comment by ProgrammingFire** [comment]
> While I think that works and doesn't look half bad, what is the reason you cannot use a combination of a repository and an enum or any arbitrary value as key, that returns some record. Why the record? Well, it's immutable and has value equality while being a reference type.
> 
> That could look like: `var subscription = repository.Get(Subscriptions.Free)` and then I can call all the method you specified on my subscription record.

You Could Do Something Like This But Now For One Feature, You Have Two Different Data Structures And You Need To Merge Which Is Kinda Bad For Clean And Clear Code.

**Comment by ufcpp** [comment]
[US Patent](https://patents.google.com/patent/US7263687B2/en)

**Comment by brianebeling** [comment]
> > While I think that works and doesn't look half bad, what is the reason you cannot use a combination of a repository and an enum or any arbitrary value as key, that returns some record. Why the record? Well, it's immutable and has value equality while being a reference type.
> > That could look like: `var subscription = repository.Get(Subscriptions.Free)` and then I can call all the method you specified on my subscription record.
> 
> You Could Do Something Like This But Now For One Feature, You Have Two Different Data Structures And You Need To Merge Which Is Kinda Bad For Clean And Clear Code.

You are right, I was kinda planning ahead there. Ignore the repository, that is not relevant to your case. @bernd5 has a great and clean solution there. If I was you I'd stick with that. In my opinion records are super clean and nice to work with.
